### PR TITLE
GAWB-2634 Add googleRole endpoints to set GCP roles

### PR DESCRIPTION
**Pull request by jmthibault79**
- Add getProxyGroup() to Sam Dao
- import workbench-model
- remove some dead code

Requires Sam and fc-dev PRs

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/resources/swagger/rawls.yaml [hunk 1]
```diff
@@ -119,7 +119,7 @@ paths:
             - openid
             - email
             - profile
-  '/api/billing/{projectId}/{role}/{email}':
+  '/api/billing/{projectId}/{workbenchRole}/{email}':
```
Comment (jmthibault79) [inline core/src/main/resources/swagger/rawls.yaml:122]: to contrast with the new `googleRole` endpoints

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 2]
```diff
@@ -150,11 +150,12 @@ object Boot extends App with LazyLogging {
     val projectOwners = gcsConfig.getStringList("projectTemplate.owners")
     val projectEditors = gcsConfig.getStringList("projectTemplate.editors")
     val projectServices = gcsConfig.getStringList("projectTemplate.services")
+    val projectOwnerGrantableRoles = gcsConfig.getStringList("projectTemplate.ownerGrantableRoles")
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:153]: not sure if `projectTemplate` is the right place

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 3]
```diff
@@ -888,12 +889,69 @@ class HttpGoogleServicesDAO(
     s"${Option(message).getOrElse("")} - code ${code}"
   }
 
+  private def addPolicyBindings(projectName: RawlsBillingProjectName, policiesToAdd: Map[String, List[String]]): Future[Unit] = {
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:892]: Refactored out of `beginProjectSetup` and also used by `addRoleToGroup`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala [hunk 4]
```diff
@@ -888,12 +889,69 @@ class HttpGoogleServicesDAO(
...
+      bindings <- retryWhen500orGoogleError(() => {
+        executeGoogleRequest(cloudResManager.projects().getIamPolicy(projectName.value, null)).getBindings
+      })
+
+      _ <- retryWhen500orGoogleError(() => {
+        val existingPolicies: Map[String, List[String]] = bindings.map { policy => policy.getRole -> policy.getMembers.toList }.toMap
+
+        // |+| is a semigroup: it combines a map's keys by combining their values' members instead of replacing them
+        import cats.implicits._
+        val newPolicies = existingPolicies |+| policiesToAdd
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:906]: Replacement for `updateBindings`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 5]
```diff
@@ -44,18 +42,14 @@ object UserService {
...
-    new UserService(userInfo, dataSource, googleServicesDAO, gpsDAO, gpsGroupSyncTopic, notificationDAO, samDAO)
+  def constructor(dataSource: SlickDataSource, googleServicesDAO: GoogleServicesDAO, gpsDAO: GooglePubSubDAO, gpsGroupSyncTopic: String, notificationDAO: NotificationDAO, samDAO: SamDAO, projectOwnerGrantableRoles: Seq[String])(userInfo: UserInfo)(implicit executionContext: ExecutionContext) =
+    new UserService(userInfo, dataSource, googleServicesDAO, gpsDAO, gpsGroupSyncTopic, notificationDAO, samDAO, projectOwnerGrantableRoles)
 
   sealed trait UserServiceMessage
   case class SetRefreshToken(token: UserRefreshToken) extends UserServiceMessage
   case object GetRefreshTokenDate extends UserServiceMessage
 
   case object CreateUser extends UserServiceMessage
-  case class AdminGetUserStatus(userRef: RawlsUserRef) extends UserServiceMessage
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:55]: These are no longer used

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 6]
```diff
@@ -237,35 +234,6 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
-  private def deleteUserFromDB(userRef: RawlsUserRef): Future[Unit] = {
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:240]: unused

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala [hunk 7]
```diff
@@ -22,28 +23,46 @@ trait BillingApiService extends HttpService with PerRequestCreator with UserInfo
...
   val billingRoutes = requireUserInfo() { userInfo =>
-    path("billing" / Segment / "members") { projectId =>
-      get {
-        requestContext =>
-          perRequest(requestContext,
-            UserService.props(userServiceConstructor, userInfo),
-            UserService.GetBillingProjectMembers(RawlsBillingProjectName(projectId)))
-      }
-    } ~
-      path("billing" / Segment / Segment / Segment) { (projectId, role, userEmail) =>
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiService.scala:33]: Moved below `googleRole` endpoints so they will match first

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala [hunk 8]
```diff
@@ -115,7 +115,7 @@ class BillingApiServiceSpec extends ApiServiceSpec {
   }
 
   it should "return 403 when adding a user to a nonexistent project" in withTestDataApiServices { services =>
-    Put(s"/billing/no_access/user/${testData.userOwner.userEmail.value}") ~>
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/BillingApiServiceSpec.scala:118]: This was testing the wrong thing

**Comment by jmthibault79** [comment]
Depends on:

https://github.com/broadinstitute/sam/pull/76
https://github.com/broadinstitute/firecloud-develop/pull/920

Required by:

https://github.com/broadinstitute/firecloud-orchestration/pull/584
https://github.com/broadinstitute/firecloud-ui/pull/1108

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala [hunk 9]
```diff
@@ -384,6 +352,33 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
...
+    if (projectOwnerGrantableRoles.contains(role))
+      Future.successful(())
+    else
+      Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.Forbidden, s"Cannot alter Google role $role: not in list [${projectOwnerGrantableRoles mkString ", "}]")))
+  }
+
+  def grantGoogleRoleToUser(projectName: RawlsBillingProjectName, targetUserEmail: WorkbenchEmail, role: String): Future[PerRequestMessage] = {
+    for {
+      _ <- googleRoleWhitelistCheck(role)
+      proxyGroupEmail <- samDAO.getProxyGroup(userInfo, targetUserEmail)
```
Comment (helgridly) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:365]: what happens if someone mistakenly puts a) a pet email b) the proxy group in here?

File: core/src/main/resources/swagger/rawls.yaml [hunk 10]
```diff
@@ -207,6 +207,92 @@ paths:
...
+            $ref: '#/definitions/ErrorReport'
+        '404':
+          description: User not found
+          schema:
+            $ref: '#/definitions/ErrorReport'
+        '500':
+          description: Rawls Internal Error
+          schema:
+            $ref: '#/definitions/ErrorReport'
+      description: grant a google role for a user in the billing project the caller owns
```
Comment (helgridly) [inline core/src/main/resources/swagger/rawls.yaml]: You probably want to mention that their pet gets access too.

**Review by helgridly — APPROVED** [review]
👍 , couple minor questions / comments

File: core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala (hunk 9 above)
Comment (jmthibault79) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:365]: Sam looks up the user subject ID for the email, so Proxy Group and Pet would both get 404

File: core/src/main/resources/swagger/rawls.yaml (hunk 10 above)
Comment (jmthibault79) [reply core/src/main/resources/swagger/rawls.yaml]: done

**Review by vkumra-broad — APPROVED** [review]

**Pull request merged**